relayer/submit_transaction_test.go

```diff
@@ -0,0 +1,119 @@
+	defer cancel()
+
+	ec := extethclient.CreateTestClient(t, sk)
+	txOpts, err := ec.TxOpts(ctx)
+	require.NoError(t, err)
+
+	// generate claim secret and public key
+	dleq := &dleq.DefaultDLEq{}
+	proof, err := dleq.Prove()
+	require.NoError(t, err)
+	res, err := dleq.Verify(proof)
+	require.NoError(t, err)
+
+	// hash public key of claim secret
+	cmt := res.Secp256k1PublicKey().Keccak256()
+
+	pub := sk.Public().(*ecdsa.PublicKey)
+	addr := crypto.PubkeyToAddress(*pub)
+
+	swapCreatorAddr, forwarderAddr := deployContracts(t, ec.Raw(), sk)
+
+	swapCreator, err := contracts.NewSwapCreator(swapCreatorAddr, ec.Raw())
+	require.NoError(t, err)
+
+	testT0Timeout := big.NewInt(300) // 5 minutes
+	testT1Timeout := testT0Timeout
+
+	value := big.NewInt(9e16)
+	nonce := big.NewInt(0)
+	txOpts.Value = value
+
+	tx, err := swapCreator.NewSwap(txOpts, cmt, [32]byte{}, addr,
+		testT0Timeout, testT1Timeout, types.EthAssetETH.Address(), value, nonce)
+	require.NoError(t, err)
+	receipt, err := block.WaitForReceipt(ctx, ec.Raw(), tx.Hash())
+	require.NoError(t, err)
+	t.Logf("gas cost to call new_swap: %d", receipt.GasUsed)
+	txOpts.Value = big.NewInt(0)
+
+	logIndex := 0 // change to 2 for ERC20, but ERC20 swaps cannot use the relayer
+	require.Equal(t, logIndex+1, len(receipt.Logs))
+	id, err := contracts.GetIDFromLog(receipt.Logs[logIndex])
+	require.NoError(t, err)
+
+	t0, t1, err := contracts.GetTimeoutsFromLog(receipt.Logs[logIndex])
+	require.NoError(t, err)
+
+	swap := &contracts.SwapCreatorSwap{
+		Owner:        addr,
+		Claimer:      addr,
+		PubKeyClaim:  cmt,
+		PubKeyRefund: [32]byte{},
+		Timeout0:     t0,
+		Timeout1:     t1,
+		Asset:        types.EthAssetETH.Address(),
+		Value:        value,
+		Nonce:        nonce,
+	}
+
+	// set contract to Ready
+	tx, err = swapCreator.SetReady(txOpts, *swap)
+	require.NoError(t, err)
+	receipt, err = block.WaitForReceipt(ctx, ec.Raw(), tx.Hash())
+	t.Logf("gas cost to call SetReady: %d", receipt.GasUsed)
+	require.NoError(t, err)
+
+	secret := proof.Secret()
+
+	// now let's try to claim
+	req, err := CreateRelayClaimRequest(ctx, sk, ec.Raw(), swapCreatorAddr, forwarderAddr, swap, &secret)
+	require.NoError(t, err)
+
+	resp, err := ValidateAndSendTransaction(ctx, req, ec, swapCreatorAddr)
+	require.NoError(t, err)
+
+	receipt, err = block.WaitForReceipt(ctx, ec.Raw(), resp.TxHash)
+	require.NoError(t, err)
+	t.Logf("gas cost to call Claim via relayer: %d", receipt.GasUsed)
+
+	// expected 1 Claimed log (ERC20 swaps have 3, but we don't support relaying with ERC20 swaps)
+	require.Equal(t, 1, len(receipt.Logs))
+
+	stage, err := swapCreator.Swaps(nil, id)
+	require.NoError(t, err)
+	require.Equal(t, contracts.StageCompleted, stage)
+
+	// Now lets try to claim a second time and verify that we fail on the simulated
+	// execution.
+	req, err = CreateRelayClaimRequest(ctx, sk, ec.Raw(), swapCreatorAddr, forwarderAddr, swap, &secret)
+	require.NoError(t, err)
```

Comment:
If I didn't call `CreateRelayClaimRequest` a second time here, I was getting a signature validation error instead of the failed simulation error. I didn't have time to look into it, but I do not believe that it is the code that checks the signature before executing. It might be good to know what code is modifying the signature and if it really should be.

Reply:
Looks like we want/need to recreate a signature a second time because the nonce is used to generate it. That's why Forwarder.sol tells us there's a signature mismatch.